File: starlore-back/src/main/java/com/robin/blogback/observability/AgentMetrics.java
package com.robin.blogback.observability;

import org.springframework.stereotype.Component;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Agent 执行指标统计。
 * <p>
 * 追踪：Token 消耗、节点耗时、工具调用频率、重试次数等。
 * 数据保留在内存中，应用重启后重置。
 */
@Component
public class AgentMetrics {

    // ── 全局计数器 ──
    private final AtomicInteger totalExecutions = new AtomicInteger(0);
    private final AtomicInteger totalRetries = new AtomicInteger(0);
    private final AtomicInteger totalErrors = new AtomicInteger(0);
    private final AtomicLong totalTokensIn = new AtomicLong(0);
    private final AtomicLong totalTokensOut = new AtomicLong(0);
    private final AtomicLong totalLatencyMs = new AtomicLong(0);

    // ── 节点耗时 ──
    private final ConcurrentHashMap<String, List<Long>> nodeLatencies = new ConcurrentHashMap<>();

    // ── 工具调用频率 ──
    private final ConcurrentHashMap<String, AtomicInteger> toolCallCounts = new ConcurrentHashMap<>();

    // ── Reviewer 决策分布 ──
    private final ConcurrentHashMap<String, AtomicInteger> reviewDecisions = new ConcurrentHashMap<>();

    // ── 审查通过率（滑动窗口） ──
    private final List<Boolean> recentPasses = Collections.synchronizedList(new ArrayList<>());
    private static final int WINDOW_SIZE = 100;

    /**
     * 记录一次完整的 Graph 执行。
     */
    public void recordExecution(int tokensIn, int tokensOut, long latencyMs,
                                 Map<String, Long> nodeTimings, int retryCount, String reviewDecision) {
        totalExecutions.incrementAndGet();
        totalTokensIn.addAndGet(tokensIn);
        totalTokensOut.addAndGet(tokensOut);
        totalLatencyMs.addAndGet(latencyMs);
        totalRetries.addAndGet(retryCount);

        // 节点耗时
        if (nodeTimings != null) {
            for (Map.Entry<String, Long> entry : nodeTimings.entrySet()) {
                nodeLatencies.computeIfAbsent(entry.getKey(), k -> Collections.synchronizedList(new ArrayList<>()))
                        .add(entry.getValue());
            }
        }

        // Reviewer 决策
        if (reviewDecision != null) {
            reviewDecisions.computeIfAbsent(reviewDecision, k -> new AtomicInteger(0)).incrementAndGet();
            recentPasses.add("PASS".equals(reviewDecision));
            // 维护滑动窗口
            while (recentPasses.size() > WINDOW_SIZE) {
                recentPasses.remove(0);
            }
        }
    }

    /**
     * 记录一次工具调用。
     */
    public void recordToolCall(String toolName) {
        toolCallCounts.computeIfAbsent(toolName, k -> new AtomicInteger(0)).incrementAndGet();
    }

    /**
     * 记录一次错误。
     */
    public void recordError() {
        totalErrors.incrementAndGet();
    }

    /**
     * 获取指标快照。
     */
    public Map<String, Object> getSnapshot() {
        Map<String, Object> snapshot = new LinkedHashMap<>();
        snapshot.put("totalExecutions", totalExecutions.get());
        snapshot.put("totalErrors", totalErrors.get());
        snapshot.put("totalRetries", totalRetries.get());
        snapshot.put("totalTokensIn", totalTokensIn.get());
        snapshot.put("totalTokensOut", totalTokensOut.get());
        snapshot.put("avgLatencyMs", totalExecutions.get() > 0
                ? totalLatencyMs.get() / totalExecutions.get() : 0);

        // 节点平均耗时
        Map<String, Long> avgNodeLatency = new LinkedHashMap<>();
        for (Map.Entry<String, List<Long>> entry : nodeLatencies.entrySet()) {
            List<Long> latencies = entry.getValue();
            avgNodeLatency.put(entry.getKey(),
                    latencies.stream().mapToLong(Long::longValue).sum() / latencies.size());
        }
        snapshot.put("avgNodeLatencyMs", avgNodeLatency);

        // 工具调用统计
        Map<String, Integer> toolStats = new LinkedHashMap<>();
        toolCallCounts.forEach((k, v) -> toolStats.put(k, v.get()));
        snapshot.put("toolCallCounts", toolStats);

        // Reviewer 决策分布
        Map<String, Integer> reviewStats = new LinkedHashMap<>();
        reviewDecisions.forEach((k, v) -> reviewStats.put(k, v.get()));
        snapshot.put("reviewDecisions", reviewStats);

        // 审查通过率
        if (!recentPasses.isEmpty()) {
            long passCount = recentPasses.stream().filter(Boolean::booleanValue).count();
            snapshot.put("passRate", Math.round(passCount * 100.0 / recentPasses.size()) + "%");
        } else {
            snapshot.put("passRate", "N/A");
        }

        return snapshot;
    }

    /**
     * 重置所有指标。
     */
    public void reset() {
        totalExecutions.set(0);
        totalRetries.set(0);
        totalErrors.set(0);
        totalTokensIn.set(0);
        totalTokensOut.set(0);
        totalLatencyMs.set(0);
        nodeLatencies.clear();
        toolCallCounts.clear();
        reviewDecisions.clear();
        recentPasses.clear();
    }
}

File: starlore-py/app/services/agent_metrics.py
"""Agent 内存级指标统计。

跟踪：
- 总执行次数、总 Token 消耗
- 各节点耗时
- 工具调用次数
- Reviewer 决策分布（PASS/REVISE/FAIL）
- 100 元素滑动窗口通过率
"""

import logging
import threading
import time
from collections import deque

logger = logging.getLogger(__name__)

_lock = threading.Lock()

# 计数器
_total_executions: int = 0
_total_tokens_in: int = 0
_total_tokens_out: int = 0

# 各节点耗时 {node_name: total_ms}
_node_latencies: dict[str, float] = {}

# 工具调用次数 {tool_name: count}
_tool_call_counts: dict[str, int] = {}

# Reviewer 决策分布 {decision: count}
_reviewer_decisions: dict[str, int] = {"PASS": 0, "REVISE": 0, "FAIL": 0}

# 100 元素滑动窗口：True=PASS, False=FAIL/REVISE
_pass_window: deque[bool] = deque(maxlen=100)


def record_execution(
    tokens_in: int = 0,
    tokens_out: int = 0,
    node_latencies: dict[str, float] | None = None,
    tool_calls: dict[str, int] | None = None,
    review_decision: str | None = None,
) -> None:
    """记录一次 Agent 执行的指标。"""
    global _total_executions, _total_tokens_in, _total_tokens_out

    with _lock:
        _total_executions += 1
        _total_tokens_in += tokens_in
        _total_tokens_out += tokens_out

        if node_latencies:
            for node, latency in node_latencies.items():
                _node_latencies[node] = _node_latencies.get(node, 0) + latency

        if tool_calls:
            for tool_name, count in tool_calls.items():
                _tool_call_counts[tool_name] = _tool_call_counts.get(tool_name, 0) + count

        if review_decision:
            decision = review_decision.upper()
            if decision in _reviewer_decisions:
                _reviewer_decisions[decision] += 1
            _pass_window.append(decision == "PASS")


def get_metrics() -> dict:
    """获取当前指标快照。"""
    with _lock:
        pass_rate = 0.0
        if _pass_window:
            pass_rate = sum(_pass_window) / len(_pass_window)

        return {
            "total_executions": _total_executions,
            "total_tokens_in": _total_tokens_in,
            "total_tokens_out": _total_tokens_out,
            "node_latencies": dict(_node_latencies),
            "tool_call_counts": dict(_tool_call_counts),
            "reviewer_decisions": dict(_reviewer_decisions),
            "pass_rate": round(pass_rate, 4),
            "pass_window_size": len(_pass_window),
            "timestamp": time.time(),
        }
